platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileWriter.java

```diff
@@ -163,64 +229,106 @@ public synchronized long writeCopiedDataItem(final long serializedVersion, final
     /**
      * Store data item in file returning location it was stored at.
      *
-     * @param dataItem
-     * 		the data item to write
+     * @param dataItem the data item to write
      * @return the data location of written data in bytes
-     * @throws IOException
-     * 		if there was a problem appending data to file
+     * @throws IOException if there was a problem appending data to file
      */
     public synchronized long storeDataItem(final D dataItem) throws IOException {
-        // find offset for the start of this new data item, we assume we always write data in a whole number of blocks
-        final long byteOffset = writePosition;
+        // find offset for the start of this new data item, we assume we always write data in a
+        // whole number of blocks
+        final int currentWritingMmapPos = writingMmap.position();
+        final long byteOffset = mmapPositionInFile + currentWritingMmapPos;
         // write serialized data
-        final int totalDataWritten = dataItemSerializer.serialize(dataItem, writingStream);
-        // update write position
-        writePosition += totalDataWritten;
+        writeWithRetry(() -> dataItemSerializer.serialize(dataItem, writingMmap), null);
         // increment data item counter
         dataItemCount++;
         // return the offset where we wrote the data
         return DataFileCommon.dataLocation(index, byteOffset);
     }
 
+    /** A helper method to write a byte buffer to the file. */
+    private void writeBytes(final ByteBuffer data) throws IOException {
+        final int pos = data.position();
+        final int limit = data.limit();
+        writeWithRetry(() -> writingMmap.put(data), () -> {
+            data.position(pos);
+            data.limit(limit);
+        });
+    }
+
     /**
-     * When you finished append to a new file, call this to seal the file and make it read only for reading.
+     * Runs a specified action, which supposed to write some data to the mapped byte buffer. If writing
+     * fails with an overflow exception, the byte buffer is remapped to the current writing position in
+     * the file channel, and the action is retried. If an additional action to reset the state is
+     * provided, it's run after the buffer is remapped, but before the writing action is executed.
      *
-     * @throws IOException
-     * 		if there was a problem sealing file or opening again as read only
+     * @param writeAction Action to run
+     * @param resetAction Action to reset the state before retry. May be null
+     * @throws IOException If an I/O error occurred
      */
-    public synchronized DataFileMetadata finishWriting() throws IOException {
+    private void writeWithRetry(final RunnableWithIOException writeAction, final RunnableWithIOException resetAction)
```

Comment:
It might be less overhead to use the functional interfaces (e.g. Consumer<ByteBuffer>) instead of runnables here (call-site syntax is about the same for either lambda implementation option, the compiler just creates an object implementing the parameter type).
Example (very approximate):
```java
writeWithRetry(data, writingMmap::put, (data) -> {...});
```
and then
```java
private void writeWithRetry(final ByteBuffer data, final Consumer<ByteBuffer> writeAction, final Consumer<ByteBuffer> resetAction) throws IOException {
...
writeAction.accept(data);
...
}
```

Reply:
It would work for `writingMap::put`, but would not compile in storeDataItem() or writeCopiedDataItem(), as they need to call methods that throw IOException.

Reply:
There are some alternatives to make it clean, but with the allocation impacts, probably best to leave it as is.